sourir.ml

```diff
@@ -19,104 +21,155 @@ let () =
+          let l = pc+1 in
+          begin match VarSet.elements xs with
+            | [x] -> Printf.eprintf
+                       "line %d: Variable %s might be uninitialized.\n%!"
+                       l x
+            | xs -> Printf.eprintf
+                      "line %d: Variables {%s} might be uninitialized.\n%!"
+                      l (String.concat ", " xs)
+          end;
+        | Scope.ExtraneousVariable (xs, pc) ->
+          let l = pc+1 in
+          let func = get_fun program f in
+          let version = get_version func v in
+          let annot = match version.annotations with | Some a -> a | None -> assert(false) in
+          let annot_vars = match annot.(pc) with
+            | None | Some (AtLeastScope _) ->
+              (* we know from the exception-raising code that this cannot happen *)
+              assert (false)
+            | Some (ExactScope vars) ->
+              VarSet.elements vars |>
+              String.concat ", " |> Printf.sprintf " {%s}" in
+          begin match VarSet.elements xs with
+            | [x] -> Printf.eprintf
+                       "line %d: Variable %s is present in scope but missing \
+                       from the scope annotation%s.\n%!"
+                       l x annot_vars
+            | xs -> Printf.eprintf
+                      "line %d: Variables {%s} are present in scope \
+                       but missing from the scope annotation%s.\n%!"
+                      l (String.concat ", " xs) annot_vars
+          end;
+        | Scope.DuplicateVariable (xs, pc) ->
+          let l = pc+1 in
+          begin match VarSet.elements xs with
+            | [x] -> Printf.eprintf
+                       "line %d: Variable %s is declared more than once.\n%!"
+                       l x
+            | xs -> Printf.eprintf
+                      "line %d: Variables {%s} are declared more than once.\n%!"
+                      l (String.concat ", " xs)
+          end;
+        | Scope.IncompatibleScope (scope1, scope2, pc) ->
+          let func = get_fun program f in
+          let version = get_version func v in
+          let instrs = version.instrs in
+          Disasm.pretty_print_version stderr (v, instrs);
+          Scope.explain_incompatible_scope stderr scope1 scope2 pc;
+          flush stderr;
+        | _ -> assert(false)
         end;
         exit 1
-      | Scope.IncompatibleScope (scope1, scope2, pc) ->
-        Disasm.pretty_print_version stderr (name, instrs);
-        Scope.explain_incompatible_scope stderr scope1 scope2 pc;
-        flush stderr;
-        exit 1
-      ) program;
+      end
+    end;
+
+    begin try Check.well_formed program with
+    | Check.MissingMain ->
+      Printf.eprintf "Program is missing an explicit or implicit main function\n";
+      exit 1
+    | Check.DuplicateFunctionDeclaration f ->
+      Printf.eprintf "Duplicate function declaration %s\n" f;
+      exit 1
+    | Check.InvalidFunctionDeclaration f ->
```

Comment:
This error (and message) isn't very helpful. Why is the function invalid?

Currently, it looks like this is only raised if the main function has a non-empty parameter list. Will this exception be used in other cases?